draft-ietf-tls-dtls13.md

```diff
@@ -2016,6 +2084,136 @@ This section provides the normative protocol types and constants definitions.
 %%## ACKs
 %%## Connection ID Management
 
+# Analysis of Limits on CCM Usage {#ccm-bounds}
+
+TLS {{?TLS13}} and {{AEBounds}} do not specify limits on usage for
+AEAD_AES_128_CCM. However, any AEAD that is used with DTLS requires limits on
+use that ensure that both confidentiality and integrity are preserved. This
+section documents that analysis for AEAD_AES_128_CCM.
+
+{{?CCM-ANALYSIS=DOI.10.1007/3-540-36492-7_7}} is used as the basis of this
+analysis. The results of that analysis are used to derive usage limits that are
+based on those chosen in {{?TLS13}}.
+
+This analysis uses symbols for multiplication (*), division (/), and
+exponentiation (^), plus parentheses for establishing precedence. The following
+symbols are also used:
+
+t:
+
+: The size of the authentication tag in bits. For this cipher, t is 128.
+
+n:
+
+: The size of the block function in bits. For this cipher, n is 128.
+
+l:
+
+: The number of blocks in each packet (see below).
+
+q:
+
+: The number of genuine packets created and protected by endpoints. This value
+  is the bound on the number of packets that can be protected before updating
+  keys.
+
+v:
+
+: The number of forged packets that endpoints will accept. This value is the
+  bound on the number of forged packets that an endpoint can reject before
+  updating keys.
+
+The analysis of AEAD_AES_128_CCM relies on a count of the number of block
+operations involved in producing each message. For simplicity, and to match the
+analysis of other AEAD functions in {{AEBounds}}, this analysis assumes a
+packet length of 2^10 blocks and a packet size limit of 2^14.
+
+For AEAD_AES_128_CCM, the total number of block cipher operations is the sum
+of: the length of the associated data in blocks, the length of the ciphertext
+in blocks, the length of the plaintext in blocks, plus 1. In this analysis,
+this is simplified to a value of twice the maximum length of a record in blocks
+(that is, `2l = 2^11`). This simplification is based on the associated data
+being limited to one block.
+
+
+## Confidentiality Limits {#ccm-confidentiality}
+
+For confidentiality, Theorem 2 in {{?CCM-ANALYSIS}} establishes that an attacker
+gains a distinguishing advantage over an ideal pseudorandom permutation (PRP) of
+no more than:
+
+~~~
+(2l * q)^2 / 2^n
```

Comment:
General remark on the analysis: While its result is stated in terms of number of packets, assuming a fixed (large) size per packet, the analysis itself seems to rather involve the total number of blocks that are being protected. For some applications with very small packet sizes, the derived recommendations in terms of packet numbers will therefore unnecessarily strict. Given that the derived constraints are of practical relevance, I wonder if it's worth stating the analysis in terms of total number of blocks, and derive a less precise but safe bound in terms of the number of packets from that.

Reply:
This is something we discussed with respect to TLS as well.  Though you might get larger limits if you count blocks, doing so is not natural in (D)TLS.  If you think that a key update every 2^20-something is onerous, then you might want to count more closely, but the large number means that you don't really need to quibble about whether it is 2^24.5 or 2^29.5.

Reply:
Does it make a noteworthy difference for CCM-8, though, where the proposed integrity bound assuming full-length packets is quite low?